File: code/ex297.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    private String res;
    public String serialize(TreeNode root) {
        res = "";
        dfs(root);
        return res;
    }
    
    // DFS
    private void dfs(TreeNode root) {
        if (root == null) {
            res += "NULL" + ",";
            return;
        }
        res = res + String.valueOf(root.val) + ",";
        dfs(root.left);
        dfs(root.right);
    }

    // Decodes your encoded data to tree.
    private int p = 0;
    public TreeNode deserialize(String data) {
        return buildTree(data);
    }
    
    private TreeNode buildTree(String data) {
        String val = nextVal(data);
        if (val.equals("NULL"))    return null;
        TreeNode root = new TreeNode(Integer.parseInt(val));
        root.left = buildTree(data);
        root.right = buildTree(data);
        return root;
    }
    
    private String nextVal(String data) {
        String s = "";
        while (data.charAt(p) != ',') {
            s += data.charAt(p++);
        }
        p++;
        return s;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));

File: code/ex297.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:
    string serial = "";
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        helper1(root);
        return serial;
    }
    
    // DFS helper to get the serial
    void helper1(TreeNode* root) {
        if (root == NULL) {
            serial = serial + "#,";
            return;
        }
        serial += to_string(root->val) + ",";
        helper1(root->left);
        helper1(root->right);
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        int idx = 0;
        return helper2(data, idx);
    }
    
    // DFS helper to build the tree
    TreeNode* helper2(string& data, int& idx) {
        string next = getNext(data, idx);
        if (next == "")    return NULL;
        if (next == "#")    return NULL;
        else {
            TreeNode* root = new TreeNode(stoi(next));
            root->left = helper2(data, idx);
            root->right = helper2(data, idx);
            return root;
        }
    }
    
    string getNext(string& data, int& idx) {
        if (idx >= data.length())    return "";
        string s = "";
        while (data[idx] != ',') {
            s += data[idx++];
        }
        idx++;
        return s;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));